Fix PythonVirtualenvOperator not working with Airflow context

feluelle opened the pull request:
Primarly this PR fixes an issue with PythonVirtualenvOperator not working in conjunction with airflow context arg.
 
Relates to #8177
Relates to #8256 

But it also has a QoL improvement. You don't need to manually pass `dill` as requirement anymore if you have set `use_dill` to `True`. So it won't raise an Error if you forgot to add it.

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

cc @maganaluis

turbaszek commented on airflow/operators/python.py:
```diff
@@ -261,6 +265,14 @@ def __init__(  # pylint: disable=too-many-arguments
+            isinstance(python_callable, types.LambdaType) and python_callable.__name__ == "<lambda>"
+        ):
+            raise AirflowException('PythonVirtualenvOperator only supports functions for python_callable arg')
+        if python_version and str(python_version)[0] != str(sys.version_info[0]) and (op_args or op_kwargs):
```
```suggestion
        if python_version and str(python_version)[0] != sys.version_info.major and (op_args or op_kwargs):
```

turbaszek commented on airflow/operators/python.py:
```diff
@@ -269,140 +281,132 @@ def __init__(  # pylint: disable=too-many-arguments
+                          f"{self._code__call_script(arg_dict='arg_dict')}" \
+                          f"{self._code__write_output(filename='sys.argv[2]')}"
+            self.log.debug('Writing code to file\n %s', python_code)
+            file.write(python_code)
```
@feluelle what would you say to use jinja template? I think it would be easier to understand how the `python_code` looks like. We used a similar approach in operator generator: 
https://github.com/PolideaInternal/airflow-munchkin/blob/master/airflow_munchkin/template/discovery/operator_class.tpl

feluelle replied:
Looks good 👍 How do you use it then?

turbaszek replied:
We render the template and save it to a file.
https://github.com/PolideaInternal/airflow-munchkin/blob/2de8bff30b342f11e260dbfa8f7725c74731adda/airflow_munchkin/discovery_parser/renderers.py#L21-L25

https://github.com/PolideaInternal/airflow-munchkin/blob/2de8bff30b342f11e260dbfa8f7725c74731adda/airflow_munchkin/block_renderer/template_utils.py#L28-L32

I think being able to take a look at the whole structure of the file is quite helpful.

feluelle replied:
Thanks - will take a look.

feluelle replied:
Done @turbaszek 👍

feluelle commented on airflow/operators/python.py:
```diff
@@ -269,140 +281,132 @@ def __init__(  # pylint: disable=too-many-arguments
+            if self.op_kwargs:
+                # some items from context can't be loaded in virtual env
+                self._keep_serializable_op_kwargs()
+            print(self.op_kwargs)
```
ToRemove

turbaszek commented on airflow/utils/python_virtualenv_script.tpl:
```diff
@@ -0,0 +1,23 @@
+# Write output
+with open(sys.argv[2], "wb") as file:
+    if res:
+        {{ pickling_library }}.dump(res, file)
```
Nice!

turbaszek commented on airflow/operators/python.py:
```diff
@@ -412,140 +427,113 @@ def __init__(  # pylint: disable=too-many-arguments
+                    python_callable_source=dedent(inspect.getsource(self.python_callable))
+                )
+            )
+            self.log.debug('Writing code to file\n %s', python_code)
```
```suggestion
            self.log.debug('Writing code to file: %s', python_code)
```
WDYT?

feluelle replied:
I added `:` but the new line should be there I think.

turbaszek commented on airflow/operators/python.py:
```diff
@@ -354,14 +354,18 @@ class PythonVirtualenvOperator(PythonOperator):
     :param python_version: The Python version to run the virtualenv with. Note that
         both 2 and 2.7 are acceptable forms.
-    :type python_version: str
+    :type python_version: str or int or float
```
```suggestion
    :type python_version: Union[str, int, float]
```

Fokko commented on airflow/operators/python.py:
```diff
@@ -412,140 +427,113 @@ def __init__(  # pylint: disable=too-many-arguments
+            with open(filename, 'wb') as file:
+                self.pickling_library.dump({'args': self.op_args, 'kwargs': self.op_kwargs}, file)
+
+    def _keep_serializable_op_kwargs(self):
```
Now we're removing all kinds of keys, wouldn't it be more preferable to have an allowed-list for keys to keep? This might be more stable in the future.

feluelle replied:
Good point. 🤔

feluelle replied:
But the problem is that you can add custom `op_kwargs` which should also be serialized but you don't know how they are called in the first place. 😁 
From the `PythonOperator` (base class of `PythonVirtualenvOperator`)
```python
    def execute(self, context: Dict):
        context.update(self.op_kwargs)
        ...
        self.op_kwargs = PythonOperator.determine_op_kwargs(self.python_callable, context, len(self.op_args))

        return_value = self.execute_callable()
        ...
```
The `PythonVirtualenvOperator` can only access `self.op_kwargs` not `context` directly which contains those unserializable objects.
I would have to overwrite the `execute` so that I can change that.

WDYT? @Fokko

feluelle replied:
I think it is actually not that hard to do. :)

Fokko reviewed (approved):
Looks good, thanks @feluelle

feluelle commented on airflow/utils/python_virtualenv.py:
```diff
@@ -69,3 +72,23 @@ def prepare_virtualenv(
+        undefined=jinja2.StrictUndefined
+    )
+    template = template_env.get_template('python_virtualenv_script.jinja2')
+    content = template.render(**jinja_context)
```
I could also use `template.stream(**jinja_context).dump(file)`. So I do not need to use `open` to create and write to a file.

turbaszek commented on airflow/operators/python.py:
```diff
@@ -413,140 +463,89 @@ def __init__(  # pylint: disable=too-many-arguments
+            return 'pendulum' in self.requirements and 'lazy_object_proxy' in self.requirements
+
+        serializable_context_keys = self.BASE_SERIALIZABLE_CONTEXT_KEYS.copy()
+        if _is_airflow_env():
+            serializable_context_keys.update(self.AIRFLOW_SERIALIZABLE_CONTEXT_KEYS)
+        if _is_pendulum_env() or _is_airflow_env():
+            serializable_context_keys.update(self.PENDULUM_SERIALIZABLE_CONTEXT_KEYS)
```
```suggestion

        if _is_pendulum_env() or _is_airflow_env():
            serializable_context_keys.update(self.AIRFLOW_SERIALIZABLE_CONTEXT_KEYS)
            serializable_context_keys.update(self.PENDULUM_SERIALIZABLE_CONTEXT_KEYS)
```
Should give the same result?

feluelle replied:
No that would mean if you have installed pendulum you can also use Airflow resources which is not correct.

turbaszek replied:
Oh I see. Will this be more readable?
```python
if _is_airflow_env():
    serializable_context_keys.update(self.AIRFLOW_SERIALIZABLE_CONTEXT_KEYS)
    serializable_context_keys.update(self.PENDULUM_SERIALIZABLE_CONTEXT_KEYS)
elif _is_pendulum_env():
    serializable_context_keys.update(self.PENDULUM_SERIALIZABLE_CONTEXT_KEYS)
```
because if `_is_airflow_env()` is true then both updates are performed. The single update is done only when `_is_airflow_env() == False` and `_is_pendulum_env()==True`

feluelle replied:
I am not sure both LGTM. What do others think? @Fokko maybe. :)

turbaszek commented on airflow/utils/python_virtualenv_script.jinja2:
```diff
@@ -0,0 +1,42 @@
+# Write output
+with open(sys.argv[2], "wb") as file:
+    if res:
+        {{ pickling_library }}.dump(res, file)
```
I like it!

feluelle commented on tests/operators/test_python.py:
```diff
@@ -1222,23 +1230,42 @@ def f(templates_dict):
+            # other
+            **context
+        ):  # pylint: disable=unused-argument,too-many-arguments,too-many-locals
+            pass
```
@turbaszek what do you think of this? I am testing it by forcing to unpack the keys given.

feluelle replied:
Also if there isn't anymore I could do I would love to get this merged. :)

feluelle replied:
> I am testing it by forcing to unpack the keys given.

More like a system-test instead of unit-test now.

turbaszek replied:
Hm, would it be possible to use paramaterization of tetst?

feluelle replied:
Good point, but I think this would rather make it more complex.. of course we could save a few lines but.. as it is right now it is pretty straight forward to see what is going on. Using `parameterized.expand` would make it more complex which is not worth the lines we would save in my opinion.

feluelle replied:
Actually i think it isn't even **possible** because the function is executed in a different context so I cannot pass variables/parameters from outside into it.

feluelle replied:
What would work would be using jinja to create the tests :D but I think that is really too much.

turbaszek reviewed (approved):
LGTM 🚀 @Fokko @kaxil  ?

feluelle committed:
Fix PythonVirtualenvOperator not working with Airflow context

- automatically add dill requirement if use_dill=True
- add howto docs
- refactor

Co-authored-by: Luis Magana <maganaluis@users.noreply.github.com>
